tensorflow/contrib/tensorrt/convert/convert_nodes.cc

```diff
@@ -1875,71 +1833,105 @@ tensorflow::Status ConvertFusedBatchNorm(
-        float& combined_offset_ref = const_cast<float*>(
-            static_cast<float const*>(combined_offset_weights.GetValues()))[i];
-        combined_scale_ref = scale / sqrtf(variance + epsilon);
-        combined_offset_ref = offset - mean * combined_scale_ref;
+  }
+  //  We could technically have two weights with different shape.
+  //  that requires two addScale op, arguably less performant
+  TRT_ShapedWeights combined_scale_weights =
+      ctx.get_temp_weights_like(*ptr_shape_weights);
+  TRT_ShapedWeights combined_offset_weights =
+      ctx.get_temp_weights_like(*ptr_shape_weights);
+
+  if (parameter_type == tensorflow::DT_FLOAT) {
+    for (size_t i = 0; i < nweight; ++i) {
+      float batchnorm_data[4];
+      for (int j = 0; j < 4; j++) {
+        if (inputs.at(j + 1).weights().count() != 1) {
+          batchnorm_data[j] = (static_cast<float const*>(
+              inputs.at(j + 1).weights().GetValues()))[i];
+        } else {
+          batchnorm_data[j] = (static_cast<float const*>(
+              inputs.at(j + 1).weights().GetValues()))[0];
+        }
       }
-    } else {
-      const Eigen::half* scale_vals =
-          (static_cast<Eigen::half const*>(scale_weights.GetValues()));
-      const Eigen::half* off_vals =
-          (static_cast<Eigen::half const*>(offset_weights.GetValues()));
-      const Eigen::half* mean_vals =
-          (static_cast<Eigen::half const*>(mean_weights.GetValues()));
-      const Eigen::half* variance_vals =
-          (static_cast<Eigen::half const*>(variance_weights.GetValues()));
-      Eigen::half* comb_scale_vals = const_cast<Eigen::half*>(
-          static_cast<Eigen::half const*>(combined_scale_weights.GetValues()));
-      Eigen::half* comb_off_vals = const_cast<Eigen::half*>(
-          static_cast<Eigen::half const*>(combined_offset_weights.GetValues()));
-      for (size_t i = 0; i < nweight; ++i) {
-        float scale(scale_vals[i]);
-        float offset(off_vals[i]);
-        float mean(mean_vals[i]);
-        float variance(variance_vals[i]);
-        float combined_scale_ref = scale / sqrtf(variance + epsilon);
-        comb_scale_vals[i] = Eigen::half(combined_scale_ref);
-        float combined_offset_ref = offset - mean * combined_scale_ref;
-        comb_off_vals[i] = Eigen::half(combined_offset_ref);
+      float scale = batchnorm_data[0];
+      float offset = batchnorm_data[1];
+      float mean = batchnorm_data[2];
+      float variance = batchnorm_data[3];
+      float& combined_scale_ref = const_cast<float*>(
+          static_cast<float const*>(combined_scale_weights.GetValues()))[i];
+      float& combined_offset_ref = const_cast<float*>(
+          static_cast<float const*>(combined_offset_weights.GetValues()))[i];
+      combined_scale_ref = scale / sqrtf(variance + epsilon);
+      combined_offset_ref = offset - mean * combined_scale_ref;
+    }
+  } else {
+    const Eigen::half* cast_vals_array[4];
```

Comment:
I think this branch (starting from line 1908) can reuse code from line 1876 to 1897 in the `if` branch, by putting the `parameter_type` check inside the `for (size_t i..)` loop? `cast_vals_array` can be of float type as it gets assigned to `batchnorm_data` which is later used as float (line 1917)?